<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CASCAM">
<title>Introduction to CASCAM package • CASCAM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to CASCAM package">
<meta property="og:description" content="CASCAM">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CASCAM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/instruction.html">Introduction to CASCAM package</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to CASCAM package</h1>
            
      
      
      <div class="d-none name"><code>instruction.Rmd</code></div>
    </div>

    
    
<p>This package is written around a S4 object <code>CASCAM</code>, where
the analysis results including the figures are stored. For now, our
framework is only limited to the bulk RNA-Seq data, and it is a
three-step framework which should be run sequentially.</p>
<ul>
<li>Step 1: Data preparation</li>
<li>Step 2: Genome-wide cancer model pre-selection</li>
<li>Step 3: Pathway specific exploration</li>
</ul>
<p>In this tutorial, we use breast cancer (BRCA) analysis as an example.
There are two breast cancer histology subtypes – invasive lobular
carcinoma (ILC) and invasive (infiltrating) ductal carcinoma (IDC), and
ILC is of interest in this study.</p>
<div class="section level2">
<h2 id="data-preparation">Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<p>In the data preparation step, the following data is necessary –</p>
<ul>
<li>
<code>tumor_log_tpm</code>: A data set of tumor gene expression on
log scale (log2-TPM preferred). It is highly recommend to merge the
interested tumor samples with the pan-cancer dataset (such as TCGA) to
benefit from the large sample size, which has shown the outperformance
comparing with the small sample size.</li>
<li>
<code>tumor_label</code>: A vector of sample labels of the tumor log
tpm data (binary class, interested v.s. uninterested).</li>
<li>
<code>camod_log_tpm</code>: A data set of cancer model gene
expression on log scale (log2-TPM preferred). Similarly, better
performance can also come from the larger sample size.</li>
<li>
<code>tumor_ct</code>: A data set of the un-normalized (estimated)
counts of sequencing reads or fragments from the tumor data.</li>
<li>
<code>tumor_label2</code>: A vector of sample labels of the tumor
count data (binary class, interested v.s. uninterested).</li>
</ul>
<div class="section level3">
<h3 id="tumor-and-cancer-model-alignment">Tumor and cancer model alignment<a class="anchor" aria-label="anchor" href="#tumor-and-cancer-model-alignment"></a>
</h3>
<p>Celligner is applied in our framework to allow the tumors and cancer
models directly comparable. Users can refer to <a href="https://www.nature.com/articles/s41467-020-20294-x" class="external-link">this paper</a>
for more information. To simplify the process, we use the default
parameters setting in Celligner paper. This step may take some time (~
30 min).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aligned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Celligner.html">Celligner</a></span><span class="op">(</span><span class="va">tumor_log_tpm</span>, <span class="va">camod_log_tpm</span><span class="op">)</span></span></code></pre></div>
<p><code>Celligner</code> returns a Seurat object with the aligned data
included, and the aligned data for tumor and cancer models of interest
(<code>tumor_aligned</code> and <code>camod_aligned</code>) can be
extracted.</p>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>Informative genes are then detected by using the tumor count matrix.
<code>DESeq2</code> and <code>fgsea</code> are implied in our framework
to identify the differential expression genes and the enriched pathways
(Hallmark and KEGG).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_InformativeGenes.html">create_InformativeGenes</a></span><span class="op">(</span><span class="va">tumor_ct</span>, <span class="va">tumor_label2</span>, <span class="st">"ILC"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-cascam-object">Create CASCAM object<a class="anchor" aria-label="anchor" href="#create-cascam-object"></a>
</h3>
<p>A <code>CASCAM</code> is then created for the downstream
analysis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CASCAM_eg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_CASCAM.html">create_CASCAM</a></span><span class="op">(</span><span class="va">tumor_aligned</span>, <span class="va">tumor_label</span>, <span class="va">camod_aligned</span>, <span class="va">gene_info</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="genome-wide-cancer-model-pre-selection">Genome-wide cancer model pre-selection<a class="anchor" aria-label="anchor" href="#genome-wide-cancer-model-pre-selection"></a>
</h2>
<div class="section level3">
<h3 id="sparse-discriminant-analysis-sda-model-trainining">Sparse discriminant analysis (SDA) model trainining<a class="anchor" aria-label="anchor" href="#sparse-discriminant-analysis-sda-model-trainining"></a>
</h3>
<p>The first step in this section is to train the SDA model for (1) high
dimensionality; (2) data sparsity; (3) classification assignment
probabilities.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CASCAM_eg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sda_model.html">sda_model</a></span><span class="op">(</span><span class="va">CASCAM_eg</span><span class="op">)</span></span></code></pre></div>
<p>By default, 1/20 of the genes are selected for the model, and the
L2-norm parameter is pre-set to 1e-4. Users can provide their own choics
or train the model through cross-validation. Parallel computation is
available in our cross-validation function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CASCAM_eg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sda_model_cv.html">sda_model_cv</a></span><span class="op">(</span><span class="va">CASCAM_eg</span><span class="op">)</span></span></code></pre></div>
<p>Starting from here, users can store the <code>CASCAM_eg</code> object
into an RData file and call <code><a href="../reference/run_CASCAM_app.html">run_CASCAM_app()</a></code> for
interactive exploration. The next step is to calculate the
genome-related measurements.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CASCAM_eg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genome_selection.html">genome_selection</a></span><span class="op">(</span><span class="va">CASCAM_eg</span><span class="op">)</span></span>
<span><span class="va">CASCAM_eg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genome_selection_visualize.html">genome_selection_visualize</a></span><span class="op">(</span><span class="va">CASCAM_eg</span><span class="op">)</span></span></code></pre></div>
<p>Users can use SDA projected positions
<code>CASCAM_eg@genome_figure$sda_project_position</code> and the
confidence intervals <code>CASCAM_eg@genome_figure$sda_ds_ci_rank</code>
to evaluate the cancer models visually.</p>
</div>
</div>
<div class="section level2">
<h2 id="pathway-specific-exploration">Pathway specific exploration<a class="anchor" aria-label="anchor" href="#pathway-specific-exploration"></a>
</h2>
<p>Pathway specific exploration is performed based on genome-wide
pre-selected cancer models. Pathway-related measurements are at first
calculated.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CASCAM_eg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pathway_analysis.html">pathway_analysis</a></span><span class="op">(</span><span class="va">CASCAM_eg</span><span class="op">)</span></span></code></pre></div>
<p>The downstream exploration then goes as follows –</p>
<ul>
<li>
<code>pathway_congruence_heatmap</code>: Explore the performance of
selected cancer models across all the available pathways using pathway
congruence heatmap.</li>
<li>
<code>pathway_specific_heatmap</code>: Select the interested pathway
and analyze the performance of selected cancer models across the genes
within that pathway using pathway specific heatmap.</li>
<li>
<code>pathway_specific_ridgeline</code>: Determine the subset of
selected cancer models to visualize their positions for each gene within
that pathway using pathway specific ridgeline .</li>
<li>
<code>pathview_analysis</code>: Choose one cancer model and
visualize its topological structure for final decision (if the pathway
is from KEGG) using pathview.</li>
</ul>
<p>It is highly recommended to performance the exploration using the
Shiny app as it provides better experience of iterative exploration.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jian Zou.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
